examples/splinterdb_intro_example.c

```diff
@@ -0,0 +1,137 @@
+#include "splinterdb/splinterdb.h"
+
+#define DB_FILE_NAME    "splinterdb_intro_db"
+#define DB_FILE_SIZE_MB 1024 // Size of SplinterDB device; Fixed when created
+#define CACHE_SIZE_MB   64   // Size of cache; can be changed across boots
+
+/* Application declares the limit of key-sizes it intends to use */
+#define APP_MAX_KEY_SIZE ((int)100)
+
+/*
+ * -------------------------------------------------------------------------------
+ * We, intentionally, do not check or show-for error handling, as this is mostly
+ * a sample program to illustrative how-to use the APIs.
+ * -------------------------------------------------------------------------------
+ */
+int
+main()
+{
+   printf("     **** SplinterDB Basic example program ****\n\n");
+
+   // Initialize data configuration, using default key-comparison handling.
+   data_config splinter_data_cfg;
+   default_data_config_init(APP_MAX_KEY_SIZE, &splinter_data_cfg);
+
+   // Basic configuration of a SplinterDB instance
+   splinterdb_config splinterdb_cfg;
+   memset(&splinterdb_cfg, 0, sizeof(splinterdb_cfg));
+   splinterdb_cfg.filename   = DB_FILE_NAME;
+   splinterdb_cfg.disk_size  = (DB_FILE_SIZE_MB * 1024 * 1024);
+   splinterdb_cfg.cache_size = (CACHE_SIZE_MB * 1024 * 1024);
+   splinterdb_cfg.data_cfg   = &splinter_data_cfg;
+
+   splinterdb *spl_handle = NULL; // To a running SplinterDB instance
+
+   int rc = splinterdb_create(&splinterdb_cfg, &spl_handle);
+   printf("Created SplinterDB instance, dbname '%s'.\n\n", DB_FILE_NAME);
+
+   // Insert a few kv-pairs, describing properties of fruits.
+   const char *fruit = "apple";
+   const char *descr = "An apple a day keeps the doctor away!";
+   slice       key   = slice_create((size_t)strlen(fruit), fruit);
+   slice       value = slice_create((size_t)strlen(descr), descr);
+
+   rc = splinterdb_insert(spl_handle, key, value);
+   printf("Inserted key '%s'\n", fruit);
+
+   fruit = "Orange";
+   descr = "Is a good source of vitamin-C.";
+   key   = slice_create((size_t)strlen(fruit), fruit);
+   value = slice_create((size_t)strlen(descr), descr);
+   rc    = splinterdb_insert(spl_handle, key, value);
+   printf("Inserted key '%s'\n", fruit);
+
+   fruit = "Mango";
+   descr = "Mango is the king of fruits.";
+   key   = slice_create((size_t)strlen(fruit), fruit);
+   value = slice_create((size_t)strlen(descr), descr);
+   rc    = splinterdb_insert(spl_handle, key, value);
+   printf("Inserted key '%s'\n", fruit);
+
+   // Retrieve a key-value pair.
+   splinterdb_lookup_result result;
+   splinterdb_lookup_result_init(spl_handle, &result, 0, NULL);
+
+   fruit = "Orange";
+   key   = slice_create((size_t)strlen(fruit), fruit);
+   rc    = splinterdb_lookup(spl_handle, key, &result);
+   rc    = splinterdb_lookup_result_value(spl_handle, &result, &value);
+   if (!rc) {
+      printf("Found key: '%s', value: '%.*s'\n",
+             fruit,
+             (int)slice_length(value),
+             (char *)slice_data(value));
+   }
```

Comment:
Is this really the way our API works now? We should really not be using the `rc` to indicate a non-error condition, such as a negative lookup.

Reply:
@ajhconway -- Currently, `splinterdb_lookup_result_value()` is implemented to return an `int`.

It calls `splinterdb_lookup_found()` defined as a `bool` function. If that were to ever fail, `splinterdb_lookup_result_value` could return a non-zero `rc`.

---

The "problem" in these interfaces is that the calling-sequence is something like this:

(Below, I added the `printf()` after `splinterdb_lookup()` to make my point.)

```
 79    rc    = splinterdb_lookup(spl_handle, key, &result);
 80    printf("splinterdb_lookup('Orange') rc=%d\n", rc);
 81    rc    = splinterdb_lookup_result_value(spl_handle, &result, &value);
 
  89    // Handling non-existent keys
 90    fruit = "Banana";
 91    key   = slice_create((size_t)strlen(fruit), fruit);
 92    rc    = splinterdb_lookup(spl_handle, key, &result);
 93    printf("splinterdb_lookup('Banana') rc=%d\n", rc);
 94    rc    = splinterdb_lookup_result_value(spl_handle, &result, &value);
 ```
 
 The output you will get when you run this is:
 
 ```
splinterdb_lookup('Orange') rc=0
Found key: 'Orange', value: 'Is a good source of vitamin-C.'
splinterdb_lookup('Banana') rc=0
Key: 'Banana' not found. (rc=22)
```

Point being: `splinterdb_lookup()` always returns 0 for both keys that exist or do not exist.
So one has to check further by calling `splinterdb_lookup_result_value`.

I agree that this is very awkward coding. I had raised similar issues in streamlining these lookup interfaces, but those were put on-hold till we get time for a revisit.

For now, I think I will leave the examples as coded till we get the time to overhaul these interfaces.